File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        out = []
        curr = []

        def dfs(start_index):
            curr_sum = sum(curr)
            if curr_sum == target:
                out.append(curr[:])
                return
            
            if start_index == len(nums) or curr_sum > target:
                return
            
            curr.append(nums[start_index])
            dfs(start_index)

            curr.pop()
            dfs(start_index + 1)

        dfs(0)
        return out

File: Data Structures & Algorithms/gas-station/submission-2.cs
public class Solution {
    public int CanCompleteCircuit(int[] gas, int[] cost) {
        int tank = 0, startIndex = 0;

        if (gas.Sum() < cost.Sum()) return -1;

        for (int i = 0; i < gas.Length; i++) {
            tank += gas[i] - cost[i];
            if (tank < 0) {
                startIndex = i + 1;
                tank = 0;
            }
        }

        return startIndex;
    }
}
